examples/run_nlp_trainer.py

```diff
@@ -0,0 +1,439 @@
+        copy = False
+        for line in infile:
+            if line.strip() == 'pretrain_hparams:':
+                copy = True
+                continue
+            elif line.strip() == 'finetune_hparams:':
+                copy = False
+                continue
+            elif copy:
+                outfile.write(line)
+
+    if args.training_scheme == 'all':  # extract run_name from trainer args for finetuning
+        hp = TrainerHparams.create(cli_args=False, f=tmp_file)
+        args.run_name = hp.run_name
+        if args.save_locally:
+            args.save_folder = hp.save_folder
+        else:
+            args.save_folder = os.path.join(args.run_name, hp.save_folder)
+
+    proc = subprocess.Popen(['composer', training_script, '-f', tmp_file, '--save_folder', args.save_folder])
+    proc.communicate()  # block until training is complete, doesn't interfere with open file streaming
+
+
+def run_finetuner(args: argparse.Namespace, finetune_hparams, glue_metrics: GlueState) -> None:
+    """Logic for handling a finetuning job spawn based on storage and training settings."""
+    # set automatic load and save paths
+    if args.training_scheme == 'all':  # load checkpoints from where pretraining saved them
+        ckpt_folder = args.save_folder
+    else:  # finetune only
+        ckpt_folder = finetune_hparams.load_path
+
+    if args.save_locally:
+        all_ckpts_list = [f for f in os.listdir(ckpt_folder) if not os.path.isdir(os.path.join(ckpt_folder, f))]
+    else:
+        all_ckpts_list = get_all_s3_checkpoints(args.bucket, ckpt_folder)
+
+    # finetune on every pretrained checkpoint
+    for ckpt_filename in all_ckpts_list:
+        parent_ckpt = ckpt_filename  # necessary for logging
+        if args.save_locally:
+            ckpt_load_path = os.path.join(ckpt_folder, ckpt_filename)
+            save_locally = True
+
+            # skip symlinks to existing checkpoints
+            if (os.path.islink(ckpt_load_path)):
+                continue
+        else:
+            ckpt_load_path = ckpt_filename
+            save_locally = False
+
+            # skip symlinks to existing checkpoints
+            if ('symlink' in ckpt_filename):
+                continue
+
+        finetune_tasks = ['cola', 'sst-2', 'qqp', 'qnli', 'mnli']
+        save_ckpts = [False, False, False, False, True]
+        setup_finetuning_jobs(finetune_tasks,
+                              save_ckpts,
+                              ckpt_load_path,
+                              ckpt_filename,
+                              ckpt_folder,
+                              glue_metrics,
+                              args.file,
+                              save_locally=save_locally,
+                              parent_ckpt=parent_ckpt,
+                              load_ignore_keys=['state/model/model.classifier*'])
+
+        # finetune on inference tasks using last mnli checkpoint
+        if args.save_locally:
```

Comment:
Yes! MNLI saves a checkpoint and then the last three finetune tasks finetune on MNLI's last checkpoint.

Reply:
Hmm ok, does that mean there is a race condition here -- if these last finetuning tasks start before the mnli job finishes?

Reply:
I dont think that condition exists because the finetune spawn calls `executor.shutdown()` which waits for all processes and callbacks to complete before returning, therefore the last 3 finetune tasks cannot start until that call finishes.